readthedocs/core/symlink.py

```diff
@@ -0,0 +1,289 @@
+"""
+A class that managed the symlinks for nginx to serve public files.
+
+All private serving will be handled by the Python app servers,
+so this only cares about public serving.
+
+The main thing that is created is a *Web Root*,
+which will be the area that is served from nginx as canonical.
+
+Considerations
+--------------
+
+* Single version docs
+
+The web root should be a symlink to the public documentation of that project.
+
+* Subprojects
+
+Subprojects will be nested under the `projects` directory,
+they will be a symlink to that project's public documentation.
+
+Objects
+-------
+
+Here is a list of all the things that the proper nouns point at:
+
+* Single Version Project -> Default Version
+* Language -> Version List
+* Subproject -> Language List
+* Project -> Language List
+
+Example layout
+--------------
+
+.. code-block::
+
+    cname_root/
+        docs.fabfile.org -> /web_root/fabric
+        pip.pypa.io -> /web_root/pip
+
+    web_root/
+        pip/
+            en/
+                latest -> rtd-builds/pip/latest/
+                stable -> rtd-builds/pip/stable/
+            ja -> /web_root/pip-ja/ja/
+            projects/
+                pip-oauth -> /web_root/pip-oauth/
+        pip-ja/
+            ja/
+
+        fabric -> rtd-builds/fabric/en/latest/ # single version
+
+"""
+
+import os
+import shutil
+import logging
+from collections import OrderedDict
+
+from django.conf import settings
+
+from readthedocs.projects.constants import LOG_TEMPLATE
+from readthedocs.projects.models import Domain
+from readthedocs.projects.utils import run
+
+log = logging.getLogger(__name__)
+
+
+class Symlink(object):
+
+    """
+    Base class for Symlinking of projects.
+
+    This will be inherited for private & public symlinking.
+    """
+
+    CNAME_ROOT = os.path.join(settings.SITE_ROOT, 'public_cname_root')
+    WEB_ROOT = os.path.join(settings.SITE_ROOT, 'public_web_root')
+    PROJECT_CNAME_ROOT = os.path.join(settings.SITE_ROOT, 'public_cname_project')
+
+    def __init__(self, project):
+        self.project = project
+        self.PROJECT_ROOT = os.path.join(
```

Comment:
Class instance variables should be lower case, upper case is reserved for constants:
https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables